app/compass_svc.py

```diff
@@ -72,3 +76,62 @@ def _get_layer_boilerplate(name, description):
             layer['techniques'].append(technique)
 
         return web.json_response(layer)
+
+    @staticmethod
+    def extract_techniques(request_body):
+        techniques = request_body.get('techniques')
+        adversary_techniques = set()
+        for technique in techniques:
+            if technique.get('score') > 0:
+                technique_id = technique.get('techniqueID')
+                adversary_techniques.add(technique_id)
+        return adversary_techniques
+
+    async def build_phases(self, adversary_techniques):
+        phases = []
+        for technique_id in adversary_techniques:
+            abilities = await self.data_svc.locate('abilities', match=dict(technique_id=technique_id))
+            for ab in abilities:
+                phases.append(dict(id=ab.ability_id, phase='1'))
+        return phases
+
+    async def read_layer(self, request):
+        chunks = []
+        reader = await request.multipart()
+        while True:
+            field = await reader.next()
+            if not field:
+                break
+            while True:
+                chunk = await field.read_chunk()
+                if not chunk:
+                    break
+                chunks.append(chunk)
+        body = b''.join(chunks)
```

Comment:
this screams python2 to me. I think there's a more python3ic way to do this part, with the byte join

Reply:
I changed to append bytes from the beginning and skip the intermediary list. But did not find a consensus on a more python3 approach. I saw PEP460: https://www.python.org/dev/peps/pep-0460/ but mentions the same approach as here